### PR TITLE
Support multiple trajectories in the asset_writer_main

### LINKED ISSUE #343: Support multi trajectory
googlecartographer/cartographer_ros/#335 started making the node multi trajectory aware. This leaks some problematic abstractions:

1) When a trajectory is finished, the node writes some derived assets (Trajectory + OccupancyGrid or X-Rays + PLY) and starts a new one. The assets use `trajectory_builder_options` for laser insertion options and resolution (which one should this be for multi trajectory?). The trajectory proto only contains the current opinion of one trajectory.
2) BuildOccupancyGrid2D - which is used to publish a legacy OccupancyGrid for ROS.

I think we have to rethink what we want to do here. I think writing the assets is fundamentally wrong in the node - instead, it should only dump the sparse pose graph (with or without submaps, as desired) and all further asset generation should be deferred to assets_writer_main. This would allow us to hold much less data in memory while slamming. 

This would also mean to pull OccupancyGrid generation into a separate ROS node that translates between Cartographer's representation and legacy ROS Occupancy grids. I am unclear if this node should keep laser data around or just reuse the submaps from cartographer and draw them on-top of each other (similar to what the visualization currently does).

/cc @macmason @wohe @yutakaoka Thoughts?

### PR DESCRIPTION
**Pull request by SirVer**
- Also track googlecartographer/cartographer#360.

Related to #343.

### CODE REVIEW

**Comment by SirVer** [comment]
Depends on #394 being merged first.

**Comment by SirVer** [comment]
Rebased and ready for review.

File: cartographer_ros/cartographer_ros/split_string.cc [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+#include "cartographer_ros/split_string.h"
+
+#include <sstream>
+
+namespace cartographer_ros {
+
+std::vector<std::string> SplitString(const std::string& input,
+                                     const char delimiter) {
+  std::stringstream stream(input);
```
Comment (wohe) [inline cartographer_ros/cartographer_ros/split_string.cc]: `std::istringstream`?

File: cartographer_ros/cartographer_ros/assets_writer_main.cc [hunk 2]
```diff
@@ -149,69 +151,82 @@ void Run(const string& pose_graph_filename, const string& bag_filename,
...
-  CHECK_EQ(pose_graph_proto.trajectory_size(), 1)
-      << "Only pose graphs containing a single trajectory are supported.";
-  const carto::mapping::proto::Trajectory& trajectory_proto =
-      pose_graph_proto.trajectory(0);
-  CHECK_GT(trajectory_proto.node_size(), 0)
-      << "Trajectory does not contain any nodes.";
+  CHECK_EQ(pose_graph_proto.trajectory_size(), bag_filenames.size())
+      << "Pose graphs contains " << pose_graph_proto.trajectory_size()
+      << " trajectories while " << bag_filenames.size()
+      << " bags where provided. This tool requires one bag for each "
```
Comment (wohe) [inline cartographer_ros/cartographer_ros/assets_writer_main.cc]: typo: `were provided`

File: cartographer_ros/cartographer_ros/assets_writer_main.cc [hunk 3]
```diff
@@ -149,69 +151,82 @@ void Run(const string& pose_graph_filename, const string& bag_filename,
...
+      const string& bag_filename = bag_filenames[trajectory_id];
+      LOG(INFO) << "Processing " << bag_filename << "...";
+      if (trajectory_proto.node_size() == 0) {
+        continue;
       }
-      if (message.isType<sensor_msgs::MultiEchoLaserScan>()) {
-        HandleMessage(*message.instantiate<sensor_msgs::MultiEchoLaserScan>(),
-                      tracking_frame, tf_buffer,
-                      *transform_interpolation_buffer, pipeline);
+      tf2_ros::Buffer tf_buffer(::ros::DURATION_MAX);
```
Comment (wohe) [inline cartographer_ros/cartographer_ros/assets_writer_main.cc]: Should there be a `TODO` that this is known to be very inefficient and should be fixed?

Comment (SirVer) [inline cartographer_ros/cartographer_ros/assets_writer_main.cc]: already further up in this file.

**Pull request merged**